File: src/main/java/ch/obermuhlner/astro/gradient/points/NearestPointsFinder.java
package ch.obermuhlner.astro.gradient.points;

import ch.obermuhlner.astro.gradient.Point;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;
import java.util.List;

public class NearestPointsFinder implements PointsFinder {

  private final int n;
  private final List<Point> points = new ArrayList<>();

  public NearestPointsFinder(int n) {
    this.n = n;
  }

  @Override
  public void setFixPoints(Collection<Point> points) {
    this.points.clear();
    this.points.addAll(points);
  }

  @Override
  public List<Point> getRelevantFixPoints(Point point) {
    points.sort(Comparator.comparingDouble(p -> point.distanceSquare(p)));

    List result = new ArrayList(n);
    for (int i = 0; i < Math.min(n, points.size()); i++) {
      result.add(points.get(i));
    }
    return result;
  }
}

File: src/main/java/ch/obermuhlner/astro/gradient/points/NearestPointsFinder.kt
package ch.obermuhlner.astro.gradient.points

import ch.obermuhlner.astro.gradient.Point
import java.util.*
import kotlin.math.min

class NearestPointsFinder constructor(private val n: Int) : PointsFinder {
    private val points: MutableList<Point> = ArrayList()

    override fun setFixPoints(points: Collection<Point>) {
        this.points.clear()
        this.points.addAll(points)
    }

    override fun getRelevantFixPoints(point: Point): List<Point> {
        points.sortWith(Comparator.comparingDouble { p: Point? -> point.distanceSquare(p) })
        val result: MutableList<Point> = ArrayList(n)
        for (i in 0 until min(n, points.size)) {
            result.add(points[i])
        }
        return result
    }
}